config/src/main/java/org/springframework/security/config/web/util/matcher/RequestMatcherBuilders.java

```diff
@@ -0,0 +1,126 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.config.web.util.matcher;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.context.ApplicationContext;
+import org.springframework.security.config.annotation.web.AbstractRequestMatcherRegistry;
+import org.springframework.security.web.util.matcher.RequestMatcherBuilder;
+import org.springframework.util.ClassUtils;
+
+/**
+ * A factory for constructing {@link RequestMatcherBuilder} instances
+ *
+ * @author Josh Cummings
+ * @since 6.2
+ */
+public final class RequestMatcherBuilders {
+
+	private static final String HANDLER_MAPPING_INTROSPECTOR = "org.springframework.web.servlet.handler.HandlerMappingIntrospector";
+
+	private static final boolean mvcPresent;
+
+	static {
+		mvcPresent = ClassUtils.isPresent(HANDLER_MAPPING_INTROSPECTOR,
+				AbstractRequestMatcherRegistry.class.getClassLoader());
+	}
+
+	private static final Log logger = LogFactory.getLog(ServletRequestMatcherBuilderFactory.class);
+
+	private RequestMatcherBuilders() {
+
+	}
+
+	/**
+	 * Create the default {@link RequestMatcherBuilder} for use by Spring Security DSLs.
+	 *
+	 * <p>If Spring MVC is not present on the classpath or if there is no {@link org.springframework.web.servlet.DispatcherServlet},
+	 * this method will return an Ant-based builder.
+	 *
+	 * <p>If the servlet configuration has only {@link org.springframework.web.servlet.DispatcherServlet}
+	 * with a single mapping (for example `/` or `/path/*`), then this method will return an MVC-based builder.
+	 *
+	 * <p>If the servlet configuration has multiple servlets, but still only one {@link org.springframework.web.servlet.DispatcherServlet}
+	 * with a single mapping (for example `/` or `/path/*`), then this method will return a builder that determines
+	 * the request matcher at request time.
+	 *
+	 * <p>Otherwise (namely if there are multiple servlet mappings for {@link org.springframework.web.servlet.DispatcherServlet}),
+	 * this with throw an exception. In that case, an application should instead use {@link ServletRequestMatcherBuilderFactory}
+	 * to create builders by servlet path.
+	 *
+	 * @param context the application context
+	 * @return the appropriate {@link RequestMatcherBuilder} based on application configuration
+	 */
+	public static RequestMatcherBuilder createDefault(ApplicationContext context) {
```

Comment:
It seems like there could be some overlap between usage of this method and `ServletRequestMatcherBuilderFactory#defaultServlet()`, and it might not be entirely clear which situations require which method to be used. It would be nice to only expose one way of creating a `RequestMatcherBuilder` at first. Since this method seems most useful for the DSL internals, I wonder if this could somehow be hidden for now?